src/reducers/console.js

```diff
@@ -35,20 +33,62 @@ export default function console(stateIn, {type, payload, meta}) {
         ),
       );
     case 'EVALUATE_CONSOLE_ENTRY':
-      return payload.trim(' ') === '' ? state : state.set(
-        meta.key,
-        new ConsoleEntry({expression: payload}),
+      return payload.trim(' ') === '' ? state : state.updateIn(
+        ['history', meta.key],
+        () => new ConsoleEntry({expression: payload}),
       );
     case 'CLEAR_CONSOLE_ENTRIES':
       return initialState;
+    case 'CONSOLE_CHANGE':
+      return state.set('currentInputValue', payload.value);
     case 'CONSOLE_LOG_PRODUCED':
-      return state.set(
-        meta.key,
-        new ConsoleEntry({
+      return state.updateIn(
+        ['history', meta.key],
+        () => new ConsoleEntry({
           value: payload.value,
           evaluatedByCompiledProjectKey: payload.compiledProjectKey,
         }),
       );
+    case 'NAVIGATE_CONSOLE_HISTORY': {
+      const relevantHistory = state.history.toList().filter(
+        entry => entry.expression !== null,
+      );
+
+      const newHistoryEntryIndex = payload.direction === 'UP' ?
+        state.historyEntryIndex + 1 :
+        state.historyEntryIndex - 1;
+
+      if (newHistoryEntryIndex < 0 ||
+          newHistoryEntryIndex === relevantHistory.size) {
```

Comment:
Much as you check `<0` for the lower bound, I would check `>== relevantHistory.size` for the upper bound to make it explicit that you are enforcing a range of allowed values.

Reply:
My hesitation with doing that is that it makes it look like the index might exceed relevantHistory.size, which the code prevents. I suppose it's inconsistent the way I have it, though, since I'm checking < 0 and === size.

I found [inRange](https://lodash.com/docs/4.17.10#inRange) in the lodash docs so I'm happy to use that to say "if the index is not in the range, return the state with no updates"

Reply:
`inRange` seems like the ideal have/eat cake!

Reply:
Even better - no external method dependency! `Immutable.List.has` can tell whether or not the index I'm at is in range.